### PR TITLE
Add tests for usm allocate/free functions and usm_allocator

### PR DESCRIPTION
**Pull request by kovshikovser**
Provide tests for:
- usm::malloc and usm::aligned_alloc overloads for all sycl::usm::alloc kinds
- usm::malloc_(device/host/shared) overloads
- usm::aligned_alloc_(device/host/shared) overloads
- usm::free
- usm_allocator constructors
- usm_allocator api

This PR relies on #131 to use common file: tests/usm/usm.h
**Please review from this commit:** https://github.com/KhronosGroup/SYCL-CTS/commit/ef0f0f98cb39306d2e61d2b53dccaf3ba36a80ae

### CODE REVIEW

File: tests/usm/usm_allocate_free.h [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    }
+  }
+};
+
+/** @brief Execute test for type defined by operation
+ *  @tparam T Testing type
+ *  @tparam op Defines name and form of usm allocation function
+ */
+template <typename T, typename op>
+static void run_test(util::logger &log) {
```
Comment (ProGTX) [inline tests/usm/usm_allocate_free.h:421]:
```suggestion
static void run_usm_test(util::logger &log) {
```

**Review by ProGTX — APPROVED** [review]

File: tests/usm/usm_allocate_free.h (hunk 1 above)
Comment (kovshikovser) [reply tests/usm/usm_allocate_free.h]: Applied

File: tests/usm/usm_allocations_helper.h [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+namespace usm_alloc_help {
+using namespace sycl_cts;
+
+/** @brief Custom type to create usm_allocator with other T type
+ */
+struct custom_type {
+  int value;
+};
+
+/** @brief Kernel index to differentiate kernel names in methods
```
Comment (keryell) [inline tests/usm/usm_allocations_helper.h:21]:
```suggestion
/** @brief Kernel index to differentiate kernel names in member functions
```
sounds more C++-friendly and actually the `enum` name is about `memb_func` :-)

**Review by keryell — COMMENTED** [review]
It is not clear to me whether you test that the shared & host USM can be used on the host in normal code, outside of a kernel running on a host device.

File: tests/usm/usm_allocations_helper.h (hunk 2 above)
Comment (kovshikovser) [reply tests/usm/usm_allocations_helper.h]: > sounds more C++-friendly and actually the enum name is about memb_func :-)

Agree, thanks, fixed

> It is not clear to me whether you test that the shared & host USM can be used on the host in normal code, outside of a kernel running on a host device.

Actually **ptr** is used on host side (line 160)
https://github.com/KhronosGroup/SYCL-CTS/blob/cee881cf0b30dda339b9c2cc376fc3b2369eb7b8/tests/usm/usm_allocations_helper.h#L154-L160

and line 178
https://github.com/KhronosGroup/SYCL-CTS/blob/cee881cf0b30dda339b9c2cc376fc3b2369eb7b8/tests/usm/usm_allocations_helper.h#L176-L180

Comment (keryell) [reply tests/usm/usm_allocations_helper.h]: Ah, yes. Thanks.

**Review by keryell — APPROVED** [review]

**Pull request merged**